### DISCUSSION TITLE
Compile a janet file to a static binary executable

### DISCUSSION THREAD

**Discussion by fnaj77** [janet-lang/janet#818] [Q&A] [answered]
Hi,
I've just discovered janet and I'm really impressed by it.
I was messing around with janet/jpm and I found interesting that I can also easily compile a janet source to a single dynamically linked executable. I was wondering if there is a way to compile a static binary as well.

I've taked a look around here and on the official jpm docs but I have not found anything about it.
Sorry if this topic has been discussed elsewhere.

Thank you again for the amazing job!

**Comment by sogaiu** [comment, marked as answer]
I believe it is possible.

The following is an edited transcript of info from @andrewchambers.  Hopefully it is still valid / not broken:

> I think all that's necessary is in your `project.janet` file, you just need to add `:lflags ["-static"]`.
>
> However, in that case if your linux distro uses glibc it won't be fully static.
>
> So to get a fully static one you can use a docker container with alpine linux to do the build.
>
> That uses musl libc which can be fully statically linked.
>
> Most linux distros use glibc.
>
> For windows and macos, not sure what else is needed.
>
> The windows flags will differ
>
> macos may just work with `-static`.
>
> For example:
```
(declare-executable
  :name "hermes"
  :entry "src/hermes-main.janet"
  :lflags [;*lib-archive-lflags*
           ;(if *static-build* ["-static"] [])]
  :deps hermes-src)
```
>
> You can check if its fully static like this: `ldd ./some-bin`
>
> If it says 'not a dynamic binary' its fully self contained, if it prints a bunch of `.so` files  something is not statically linked.
>
> Your alpine docker container probably needs the 'musl-dev' and maybe 'musl-static' packages.
>
> I can't remember.
>
> You install them like `apk add gcc musl-dev`.
>
> So the caveat being, only C extensions you preload using the new preload thing will be importable from your user scripts afaik.

**Reply by fnaj77** [reply]
hi @sogaiu,
adding :lflags ["-static"] do the job because I'm already under musl, so no problem due to glibc.

Thank you for your fast and clear help!
